File: 2/lab22.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <set>

using namespace std;

string findLongestChain(const vector<pair<string, string>>& trips) {
    // Карта: из какого города -> в какой город
    map<string, string> nextCity;
    // Множество всех городов
    set<string> allCities;
    // Множество городов, куда можно приехать
    set<string> destinations;
    
    // Заполняем контейнеры
    for (const auto& trip : trips) {
        string from = trip.first;
        string to = trip.second;
        
        // Проверка на петлю (город сам в себя)
        if (from == to) {
            return "Ошибка";
        }
        
        nextCity[from] = to;
        allCities.insert(from);
        allCities.insert(to);
        destinations.insert(to);
    }
    
    // Ищем начальный город (который никуда не ведет как пункт назначения)
    string start;
    for (const auto& city : allCities) {
        if (destinations.find(city) == destinations.end()) {
            start = city;
            break;
        }
    }
    
    // Если начальный город не нашли - значит есть цикл
    if (start.empty()) {
        return "Ошибка";
    }
    
    // Идем по цепочке до конца
    string current = start;
    while (nextCity.find(current) != nextCity.end()) {
        current = nextCity[current];
    }
    
    // current - это конечный пункт
    return current;
}

int main() {
    // Пример 1
    vector<pair<string, string>> trips1 = {
        {"Лондон", "Стамбул"},
        {"Новосибирск", "Лондон"},
        {"Стамбул", "Пекин"}
    };
    cout << findLongestChain(trips1) << endl; // Пекин
    
    // Пример 2
    vector<pair<string, string>> trips2 = {
        {"Барнаул", "Новосибирск"}
    };
    cout << findLongestChain(trips2) << endl; // Новосибирск
    
    // Пример 3 (петля)
    vector<pair<string, string>> trips3 = {
        {"Искитим", "Искитим"}
    };
    cout << findLongestChain(trips3) << endl; // Ошибка
    
    // Пример 4 (цикл)
    vector<pair<string, string>> trips4 = {
        {"Москва", "Норильск"},
        {"Норильск", "Саратов"},
        {"Саратов", "Москва"}
    };
    cout << findLongestChain(trips4) << endl; // Ошибка
    
    return 0;
}

File: 2/lab22.swift
import Foundation

func findLongestChain(_ trips: [(String, String)]) -> String {
    // Карта: из какого города -> в какой город
    var nextCity: [String: String] = [:]
    // Множество всех городов
    var allCities = Set<String>()
    // Множество городов, куда можно приехать
    var destinations = Set<String>()
    
    // Заполняем контейнеры
    for trip in trips {
        let from = trip.0
        let to = trip.1
        
        // Проверка на петлю (город сам в себя)
        if from == to {
            return "Ошибка"
        }
        
        nextCity[from] = to
        allCities.insert(from)
        allCities.insert(to)
        destinations.insert(to)
    }
    
    // Ищем начальный город (который никуда не ведет как пункт назначения)
    var start = ""
    for city in allCities {
        if !destinations.contains(city) {
            start = city
            break
        }
    }
    
    // Если начальный город не нашли - значит есть цикл
    if start.isEmpty {
        return "Ошибка"
    }
    
    // Идем по цепочке до конца
    var current = start
    while nextCity.keys.contains(current) {
        current = nextCity[current]!
    }
    
    // current - это конечный пункт
    return current
}

// Пример 1
let trips1: [(String, String)] = [
    ("Лондон", "Стамбул"),
    ("Новосибирск", "Лондон"),
    ("Стамбул", "Пекин")
]
print(findLongestChain(trips1)) // Пекин

// Пример 2
let trips2: [(String, String)] = [
    ("Барнаул", "Новосибирск")
]
print(findLongestChain(trips2)) // Новосибирск

// Пример 3 (петля)
let trips3: [(String, String)] = [
    ("Искитим", "Искитим")
]
print(findLongestChain(trips3)) // Ошибка

// Пример 4 (цикл)
let trips4: [(String, String)] = [
    ("Москва", "Норильск"),
    ("Норильск", "Саратов"),
    ("Саратов", "Москва")
]
print(findLongestChain(trips4)) // Ошибка
